### PR TITLE
user validation

### LINKED ISSUE #80: user creation validations
Validate that:
- [ ] Manager
- [ ] SubManager
- [ ] DataEntryClerk
- [ ] Cashier
must be connected to a shop

validation should be put inside the clean method,
and full_clean method should be called inside the save method

### PR DESCRIPTION
**Pull request by HMMK99**

### CODE REVIEW

**Comment by mmostafa74** [comment]
run black to fix flake8 issues

**Comment by mmostafa74** [comment]
:rocket:

File: xshop/users/tests/test_models.py [hunk 1]
```diff
@@ -162,10 +180,15 @@ def test_created_general_manager_attrs(self):
         self.assertEqual(self.gm.mobile, "+201010092181")
         self.assertTrue(self.gm.is_active)
         self.assertFalse(self.gm.is_staff)
+        self.assertEqual(self.gm.shop, self.shop)
         self.assertFalse(self.gm.is_superuser)
         self.assertIsNone(self.gm.username)
         self.assertEqual(GeneralManager.objects.count(), 1)
 
+    def test_valid_err_if_noshop_with_gm(self):
+        with self.assertRaises(ValidationError):
```
Comment (shahwan42) [inline xshop/users/tests/test_models.py]: make sure the message is the same as the raised error's,
check here for more:
https://docs.djangoproject.com/en/3.1/topics/testing/tools/#assertions

File: xshop/users/tests/test_models.py [hunk 2]
```diff
@@ -137,16 +149,22 @@ def test_dec_added_to_dec_group(self):
...
         self.assertEqual(self.manager.mobile, "+201010092181")
         self.assertTrue(self.manager.is_active)
         self.assertFalse(self.manager.is_staff)
         self.assertFalse(self.manager.is_superuser)
+        self.assertEqual(self.manager.shop, self.shop)
         self.assertIsNone(self.manager.username)
         self.assertEqual(Manager.objects.count(), 1)
 
+    def test_valid_err_if_noshop_with_manager(self):
+        with self.assertRaises(ValidationError):
```
Comment (shahwan42) [inline xshop/users/tests/test_models.py]: here too

File: xshop/users/tests/test_models.py [hunk 3]
```diff
@@ -119,16 +125,22 @@ def test_cashier_added_to_cashier_group(self):
...
         self.assertEqual(self.dec.mobile, "+201010092181")
         self.assertTrue(self.dec.is_active)
         self.assertFalse(self.dec.is_staff)
         self.assertFalse(self.dec.is_superuser)
+        self.assertEqual(self.dec.shop, self.shop)
         self.assertIsNone(self.dec.username)
         self.assertEqual(DataEntryClerk.objects.count(), 1)
 
+    def test_valid_err_if_noshop_with_DEC(self):
+        with self.assertRaises(ValidationError):
```
Comment (shahwan42) [inline xshop/users/tests/test_models.py]: here too

File: xshop/users/tests/test_models.py [hunk 4]
```diff
@@ -101,16 +101,22 @@ def test_customer_added_to_customer_group(self):
...
         self.assertEqual(self.cashier.mobile, "+201010092181")
         self.assertTrue(self.cashier.is_active)
         self.assertFalse(self.cashier.is_staff)
         self.assertFalse(self.cashier.is_superuser)
+        self.assertEqual(self.cashier.shop, self.shop)
         self.assertIsNone(self.cashier.username)
         self.assertEqual(Cashier.objects.count(), 1)
 
+    def test_valid_err_if_noshop_with_Cashier(self):
+        with self.assertRaises(ValidationError):
```
Comment (shahwan42) [inline xshop/users/tests/test_models.py]: here too

**Review by shahwan42 — CHANGES_REQUESTED** [review]

File: xshop/users/mixins.py [hunk 5]
```diff
@@ -19,3 +23,14 @@ def has_delete_permission(self, request, obj=None):
     def has_add_permission(self, request, obj=None):
         if request.user.is_authenticated and request.user.is_superuser:
             return True
+
+
+class ShopRequiredMixin(object):
```
Comment (shahwan42) [inline xshop/users/mixins.py:28]: no need to specify `object` to inherit from

**Review by shahwan42 — APPROVED** [review]

**Pull request merged**